server/src/main/java/com/genymobile/scrcpy/UHidManager.java

```diff
@@ -0,0 +1,98 @@
+package com.genymobile.scrcpy;
+
+import android.annotation.TargetApi;
+import android.os.Build;
+import android.os.HandlerThread;
+import android.os.MessageQueue;
+import android.system.ErrnoException;
+import android.system.Os;
+import android.system.OsConstants;
+
+import java.io.FileDescriptor;
+import java.io.InterruptedIOException;
+import java.nio.ByteBuffer;
+import java.util.Hashtable;
+
+public class UHidManager {
+    private final Hashtable<Integer, FileDescriptor> mHandles = new Hashtable<>();
+    private final HandlerThread thread = new HandlerThread("UHidManager");
+    private final MessageQueue queue;
+    private final DeviceMessageSender sender;
+    private final ByteBuffer buffer = ByteBuffer.allocateDirect(5 * 1024);
+
+    @TargetApi(Build.VERSION_CODES.M)
+    public UHidManager(DeviceMessageSender sender) {
+        thread.start();
+        queue = thread.getLooper().getQueue();
+        this.sender = sender;
+    }
+
+    public void stop() {
+        thread.quit();
+    }
+
+    public void join() {
+        try {
+            thread.join();
+        } catch (InterruptedException e) {
+            Ln.e("Failed to join UHidManager thread: ", e);
+        }
+    }
+
+    @TargetApi(Build.VERSION_CODES.M)
+    public void open(int id) {
+        FileDescriptor fd = null;
+        try {
+            fd = Os.open("/dev/uhid", OsConstants.O_RDWR, 0);
+        } catch (ErrnoException e) {
+            Ln.e("Failed to open uhid", e);
+            return;
+        }
+
+        mHandles.put(id, fd);
+        queue.addOnFileDescriptorEventListener(fd, MessageQueue.OnFileDescriptorEventListener.EVENT_INPUT | MessageQueue.OnFileDescriptorEventListener.EVENT_OUTPUT, (fd2, event) -> {
+            if (event == MessageQueue.OnFileDescriptorEventListener.EVENT_INPUT) {
+                try {
+                    buffer.rewind();
+                    Os.read(fd2, buffer);

```

Comment:
Do you know why it reads 4380 bytes (with only the first byte being non-zero)?

That does not match `struct uhid_output_req` either ([documentation](https://raw.githubusercontent.com/torvalds/linux/master/Documentation/hid/uhid.rst)):
https://github.com/torvalds/linux/blob/603c04e27c3e9891ce7afa5cd6b496bfacff4206/include/uapi/linux/uhid.h#L81-L85

Reply:
4380 is the size of `uhid_event`. It always reads/writes full `uhid_event`s.